src/optimizer/plugins/SimplifyArrayIndexingOptimizer.ts

```diff
@@ -0,0 +1,123 @@
+        return parsedPath;
+      });
+
+      parsedRules.forEach((parsedRule: any, ruleIndex: number) => {
+        const originalRule = def.rules[ruleIndex];
+        parsedRule.path.forEach((element: fhirtypes.PathPart, elementIndex: number) => {
+          // Add a prefix to the current element containing previously parsed rule elements
+          element.prefix = utils.assembleFSHPath(parsedRule.path.slice(0, elementIndex));
+          applySoftIndexing(element, pathMap);
+        });
+
+        parsedRule.caretPath?.forEach((element: fhirtypes.PathPart, elementIndex: number) => {
+          // Caret path indexes should only be resolved in the context of a specific path
+          // Each normal path has a separate map to keep track of the caret path indexes
+          if (!caretPathMap.has(originalRule.path)) {
+            caretPathMap.set(originalRule.path, new Map());
+          }
+
+          const elementCaretPathMap = caretPathMap.get(originalRule.path);
+          // Add a prefix to the current element containing previously parsed rule elements
+          element.prefix = utils.assembleFSHPath(parsedRule.caretPath.slice(0, elementIndex));
+          applySoftIndexing(element, elementCaretPathMap);
+        });
+      });
+
+      removeZeroIndices();
+
+      parsedRules.forEach((parsedRule: any, ruleIndex: number) => {
+        const originalRule = def.rules[ruleIndex];
+        originalRule.path = utils.assembleFSHPath(parsedRule.path);
+        if (originalRule instanceof ExportableCaretValueRule) {
+          originalRule.caretPath = utils.assembleFSHPath(parsedRule.caretPath);
+        }
+      });
+    });
+  }
+} as OptimizerPlugin;
+
+/**
+ *
+ * @param {PathPart} element - A single element in a rules path
+ * @param {Map<string, number} pathMap - A map containing an element's name as the key and that element's updated index as the value
+ */
+function applySoftIndexing(element: fhirtypes.PathPart, pathMap: Map<string, number>) {
+  // Must account for a pathPart's base name, prior portions of the path, as well as any slices it's contained in.
+  const mapName = `${element.prefix ?? ''}.${element.base}|${(element.slices ?? []).join('|')}`;
+  const indexRegex = /^[0-9]+$/;
+
+  const currentNumericBracket = element.brackets?.find(bracket => indexRegex.test(bracket));
+  if (!currentNumericBracket) return; // If the element is not an array, exit the function
+
+  const bracketIndex = element.brackets.indexOf(currentNumericBracket);
+
+  if (!pathMap.has(mapName)) {
+    pathMap.set(mapName, parseInt(currentNumericBracket));
+    // If the element contains a zero Index, we'll flag the rule to have the index removed
+    if (currentNumericBracket === '0') singletonArrayElements.set(mapName, [element]);
+  } else {
+    const previousNumericIndex = pathMap.get(mapName);
+    if (parseInt(currentNumericBracket) === previousNumericIndex + 1) {
+      if (previousNumericIndex === 0) singletonArrayElements.delete(mapName);
```

Comment:
The single-line style for `if` statements is a little bit tricky to read, especially in a position where it is right after another `if` statement and followed by additional statements. Could you put braces around all of your `if`-blocks?

Reply:
Yes I can!

Reply:
Addressed in [`d6e165e`](https://github.com/FHIR/GoFSH/pull/106/commits/d6e165e5611b739e6fae61fe8ba1ca39183768ae).